src/vs/workbench/parts/files/browser/views/explorerViewer.ts

```diff
@@ -526,20 +526,53 @@ export class FileController extends DefaultController {
 
 // Explorer Sorter
 export class FileSorter implements ISorter {
+	private toDispose: IDisposable[];
+	private sortOrder: SortOrder;
+
+	constructor(
+		@IConfigurationService private configurationService: IConfigurationService
+	) {
+		this.toDispose = [];
+
+		this.onConfigurationUpdated(configurationService.getConfiguration<IFilesConfiguration>());
+
+		this.registerListeners();
+	}
+
+	private registerListeners(): void {
+		this.toDispose.push(this.configurationService.onDidUpdateConfiguration(e => this.onConfigurationUpdated(this.configurationService.getConfiguration<IFilesConfiguration>())));
+	}
+
+	private onConfigurationUpdated(configuration: IFilesConfiguration): void {
+		this.sortOrder = configuration && configuration.explorer && configuration.explorer.sortOrder || 'default';
+	}
 
 	public compare(tree: ITree, statA: FileStat, statB: FileStat): number {
-		if (statA.isDirectory && !statB.isDirectory) {
-			return -1;
-		}
+		switch (this.sortOrder) {
```

Comment:
I appreciate that we now can do this all from a single compare implementation, that is very cool. However, I am having some difficulties that we have the `switch (this.sortOrder)` twice in the method, makes it very hard to understand the flow imho. Can we:
* move the things which are independent from the sort order up (`NewStatPlaceholder` and `roots`)
* have a single switch statement and not two?

I know that having a single switch statement might mean some duplication of sorting logic, but I think it would still be easier to understand. We can still in that model extract logic into methods if we wanted to share.